# 1 Hands on

Welcome to this hands on about sorting data using pandas. This notebook presents the practical side of the concepts explained in the video. The dataset used in this notebook can be downloaded from [here](http://www.fueleconomy.gov/feg/epadata/vehicles.csv.zip).This hands on covers the following topics:
• Sorting DataFrames based on single column
• Sorting DataFrames based on multiple columns
• Sorting in ascending and descending order
• Sorting based on row and column labels
• Sorting in the presence of missing values
Let us first read the required data in a DataFrame and perform some preliminary investigation on the data.

In [2]:
import pandas as pd 

The dataset has 26 columns but for the purpose of this demo, we will use only a few columns which have been shown below as a list.

In [3]:
sub_columns = ["id", "make", "model", "year", "cylinders", "fuelType", "trany", "mpgData", "city08", "highway08"]

We will now read only these columns in the DataFrame.

In [4]:
df1= pd.read_csv('vehicles.csv', usecols= sub_columns)

In [5]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47075 entries, 0 to 47074
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   city08     47075 non-null  int64  
 1   cylinders  46423 non-null  float64
 2   fuelType   47075 non-null  object 
 3   highway08  47075 non-null  int64  
 4   id         47075 non-null  int64  
 5   make       47075 non-null  object 
 6   model      47075 non-null  object 
 7   mpgData    46839 non-null  object 
 8   trany      47064 non-null  object 
 9   year       47075 non-null  int64  
dtypes: float64(1), int64(4), object(5)
memory usage: 3.6+ MB


df1.head()

The dataset has 47074 rows and 10 columns.

# 1.1  Does the data have any null values?

In [9]:
null_vals = df1.isnull().sum()

In [10]:
print(null_vals[null_vals!=0])

cylinders    652
mpgData      236
trany         11
dtype: int64


The columns having null values and their respective counts have been displayed above.

# 1.2  Sorting data based on a column

Let’s sort df1 based on the values of column fuelType using sort_values(). This column is a stringtype column so sorting would occur from A- Z, followed by a-z. The sorted DataFrame is shownbelow.

In [27]:
df1.sort_values('fuelType')

city08  cylinders                     fuelType  highway08     id  \
17450      24        4.0                          CNG         36  25868   
9174       12        8.0                          CNG         14  18256   
9173       11        8.0                          CNG         15  18255   
9172       14        8.0                          CNG         20  18254   
5006       10        8.0                          CNG         14  14504   
...       ...        ...                          ...        ...    ...   
40977      29        4.0  Regular Gas and Electricity         29  47275   
33991      43        4.0   Regular Gas or Electricity         42  40924   
32069      43        4.0   Regular Gas or Electricity         42  39190   
29495      43        4.0   Regular Gas or Electricity         42  36863   
29983      43        4.0   Regular Gas or Electricity         42  37309   

            make                 model mpgData  \
17450      Honda             Civic CNG       Y   
9174       Dodge  Ram Van 2500 2WD CNG       N   
9173        Ford              F150 CNG       N   
9172        Ford    Crown Victoria CNG       Y   
5006        Ford   F250 Pickup 2WD CNG       N   
...          ...                   ...     ...   
40977      Dodge       Hornet PHEV AWD       N   
33991  Chevrolet                  Volt       N   
32069  Chevrolet                  Volt       Y   
29495  Chevrolet                  Volt       Y   
29983  Chevrolet                  Volt       Y   

                                  trany  year  
17450                   Automatic 5-spd  2009  
9174                    Automatic 4-spd  2002  
9173                    Automatic 4-spd  2002  
9172                    Automatic 4-spd  2002  
5006                    Automatic 4-spd  1998  
...                                 ...   ...  
40977                   Automatic 6-spd  2024  
33991  Automatic (variable gear ratios)  2019  
32069  Automatic (variable gear ratios)  2018  
29495  Automatic (variable gear ratios)  2016  
29983  Automatic (variable gear ratios)  2017  

[47075 rows x 10 columns]

The sorted does not affect the original DataFrame. It only returns a copy of the sorted DataFrame.This can be verified by looking at df1.

In [12]:
df1.head()

city08  cylinders fuelType  highway08     id        make  \
0      19        4.0  Regular         25      1  Alfa Romeo   
1       9       12.0  Regular         14     10     Ferrari   
2      23        4.0  Regular         33    100       Dodge   
3      10        8.0  Regular         12   1000       Dodge   
4      17        4.0  Premium         23  10000      Subaru   

                 model mpgData            trany  year  
0   Spider Veloce 2000       Y     Manual 5-spd  1985  
1           Testarossa       N     Manual 5-spd  1985  
2              Charger       Y     Manual 5-spd  1985  
3  B150/B250 Wagon 2WD       N  Automatic 3-spd  1985  
4     Legacy AWD Turbo       N     Manual 5-spd  1993

Let us now sort df1 by using the column city08 and see the output.

In [13]:
df1.sort_values('city08')

city08  cylinders     fuelType  highway08     id         make  \
21057       6       12.0      Premium         10   2914  Lamborghini   
7901        6       12.0      Premium         10   1711  Lamborghini   
34522       6       12.0      Premium         10   4141  Lamborghini   
43092       6       12.0      Premium         10   6413  Lamborghini   
41818       6       12.0      Premium         10   5267  Lamborghini   
...       ...        ...          ...        ...    ...          ...   
33393     150        NaN  Electricity        122  40384      Hyundai   
40625     151        NaN  Electricity        120  46959      Hyundai   
40256     151        NaN  Electricity        120  46624      Hyundai   
40254     153        NaN  Electricity        127  46622      Hyundai   
40623     153        NaN  Electricity        127  46957      Hyundai   

                                         model mpgData           trany  year  
21057                                 Countach       N    Manual 5-spd  1987  
7901                                  Countach       N    Manual 5-spd  1986  
34522                                 Countach       N    Manual 5-spd  1988  
43092                                 Countach       N    Manual 5-spd  1990  
41818                                 Countach       N    Manual 5-spd  1989  
...                                        ...     ...             ...   ...  
33393                           Ioniq Electric       N  Automatic (A1)  2019  
40625               Ioniq 6 Standard Range RWD       N  Automatic (A1)  2024  
40256               Ioniq 6 Standard Range RWD       N  Automatic (A1)  2023  
40254  Ioniq 6 Long range RWD (18 inch Wheels)       N  Automatic (A1)  2023  
40623  Ioniq 6 Long range RWD (18 inch Wheels)       N  Automatic (A1)  2024  

[47075 rows x 10 columns]

# 1.3  Sorting data in descending order

Sorting data in descending order can be done by setting ‘ascending= False’ as shown.

In [14]:
df1.sort_values(by = 'city08', ascending= False)

city08  cylinders     fuelType  highway08     id         make  \
40254     153        NaN  Electricity        127  46622      Hyundai   
40623     153        NaN  Electricity        127  46957      Hyundai   
40256     151        NaN  Electricity        120  46624      Hyundai   
40625     151        NaN  Electricity        120  46959      Hyundai   
31226     150        NaN  Electricity        122  38431      Hyundai   
...       ...        ...          ...        ...    ...          ...   
43092       6       12.0      Premium         10   6413  Lamborghini   
7901        6       12.0      Premium         10   1711  Lamborghini   
21057       6       12.0      Premium         10   2914  Lamborghini   
34522       6       12.0      Premium         10   4141  Lamborghini   
41818       6       12.0      Premium         10   5267  Lamborghini   

                                         model mpgData           trany  year  
40254  Ioniq 6 Long range RWD (18 inch Wheels)       N  Automatic (A1)  2023  
40623  Ioniq 6 Long range RWD (18 inch Wheels)       N  Automatic (A1)  2024  
40256               Ioniq 6 Standard Range RWD       N  Automatic (A1)  2023  
40625               Ioniq 6 Standard Range RWD       N  Automatic (A1)  2024  
31226                           Ioniq Electric       N  Automatic (A1)  2017  
...                                        ...     ...             ...   ...  
43092                                 Countach       N    Manual 5-spd  1990  
7901                                  Countach       N    Manual 5-spd  1986  
21057                                 Countach       N    Manual 5-spd  1987  
34522                                 Countach       N    Manual 5-spd  1988  
41818                                 Countach       N    Manual 5-spd  1989  

[47075 rows x 10 columns]

# 1.4  Sorting using multiple columns

Sorting can also be done using multiple columns as shown below. Observe the last 3 rows of the below output to see the effect of using multiple columns for sorting.

In [15]:
df1.sort_values(by=['city08', 'highway08']).head(20)

city08  cylinders fuelType  highway08     id          make  \
7901        6       12.0  Premium         10   1711   Lamborghini   
21057       6       12.0  Premium         10   2914   Lamborghini   
34522       6       12.0  Premium         10   4141   Lamborghini   
41818       6       12.0  Premium         10   5267   Lamborghini   
43092       6       12.0  Premium         10   6413   Lamborghini   
1979        7        8.0  Regular          9   1178  Aston Martin   
1990        7        8.0  Regular          9   1179  Aston Martin   
13641       7        8.0  Regular          9   2228   Rolls-Royce   
18792       7        8.0  Regular          9  27081   Rolls-Royce   
19947       7        8.0  Regular          9  28120   Rolls-Royce   
23166       7        8.0  Regular          9   3110   Rolls-Royce   
23199       7        8.0  Regular          9   3113   Rolls-Royce   
23221       7        8.0  Regular          9   3115   Rolls-Royce   
25346       7        8.0  Regular          9   3308   Rolls-Royce   
27193       7        8.0  Regular          9   3477   Rolls-Royce   
40670       7        8.0  Regular          9     47  Aston Martin   
41298       7        8.0  Regular          9     48  Aston Martin   
42287       7        8.0  Regular          9    569   Rolls-Royce   
1770        7       12.0  Premium         10  11591       Ferrari   
2002        7        8.0  Regular         10   1180  Aston Martin   

                        model mpgData            trany  year  
7901                 Countach       N     Manual 5-spd  1986  
21057                Countach       N     Manual 5-spd  1987  
34522                Countach       N     Manual 5-spd  1988  
41818                Countach       N     Manual 5-spd  1989  
43092                Countach       N     Manual 5-spd  1990  
1979                  Lagonda       N  Automatic 3-spd  1985  
1990   Saloon/Vantage/Volante       N  Automatic 3-spd  1985  
13641   Silver Spur Limousine       N  Automatic 3-spd  1986  
18792   Silver Spur Limousine       N  Automatic 3-spd  1984  
19947   Silver Spur Limousine       N  Automatic 3-spd  1984  
23166             Continental       N  Automatic 3-spd  1987  
23199             Corniche II       N  Automatic 3-spd  1987  
23221             Corniche II       N  Automatic 3-spd  1987  
25346                Camargue       N  Automatic 3-spd  1987  
27193   Silver Spur Limousine       N  Automatic 3-spd  1987  
40670                 Lagonda       N  Automatic 3-spd  1985  
41298  Saloon/Vantage/Volante       N  Automatic 3-spd  1985  
42287   Silver Spur Limousine       N  Automatic 3-spd  1985  
1770              Ferrari F50       N     Manual 6-spd  1995  
2002   Saloon/Vantage/Volante       N     Manual 5-spd  1985

# 2  Sorting multiple columns in descending order

In [16]:
df1.sort_values(by=['city08','highway08'], ascending= False).head(20)

city08  cylinders     fuelType  highway08     id     make  \
40254     153        NaN  Electricity        127  46622  Hyundai   
40623     153        NaN  Electricity        127  46957  Hyundai   
40256     151        NaN  Electricity        120  46624  Hyundai   
40625     151        NaN  Electricity        120  46959  Hyundai   
37173     150        NaN  Electricity        133  43821    Tesla   
31226     150        NaN  Electricity        122  38431  Hyundai   
32569     150        NaN  Electricity        122  39641  Hyundai   
33393     150        NaN  Electricity        122  40384  Hyundai   
35477     148        NaN  Electricity        132  42278    Tesla   
35472     145        NaN  Electricity        121  42273  Hyundai   
37232     145        NaN  Electricity        121  43879  Hyundai   
39906     141        NaN  Electricity        140  46307    Lucid   
39908     141        NaN  Electricity        140  46309    Lucid   
36711     141        NaN  Electricity        127  43401    Tesla   
41171     140        NaN  Electricity        134  47454    Lucid   
34529     140        NaN  Electricity        124  41416    Tesla   
37234     140        NaN  Electricity        119  43880    Tesla   
38476     140        NaN  Electricity        119  45017    Tesla   
38472     138        NaN  Electricity        126  45013    Tesla   
39794     138        NaN  Electricity        126  46206    Tesla   

                                         model mpgData           trany  year  
40254  Ioniq 6 Long range RWD (18 inch Wheels)       N  Automatic (A1)  2023  
40623  Ioniq 6 Long range RWD (18 inch Wheels)       N  Automatic (A1)  2024  
40256               Ioniq 6 Standard Range RWD       N  Automatic (A1)  2023  
40625               Ioniq 6 Standard Range RWD       N  Automatic (A1)  2024  
37173          Model 3 Standard Range Plus RWD       Y  Automatic (A1)  2021  
31226                           Ioniq Electric       N  Automatic (A1)  2017  
32569                           Ioniq Electric       N  Automatic (A1)  2018  
33393                           Ioniq Electric       N  Automatic (A1)  2019  
35477              Model 3 Standard Range Plus       N  Automatic (A1)  2020  
35472                           Ioniq Electric       N  Automatic (A1)  2020  
37232                           Ioniq Electric       Y  Automatic (A1)  2021  
39906         Air Pure AWD with 19 inch wheels       N  Automatic (A1)  2023  
39908      Air Touring AWD with 19 inch wheels       N  Automatic (A1)  2023  
36711                   Model 3 Long Range AWD       Y  Automatic (A1)  2021  
41171         Air Pure RWD with 19 inch wheels     NaN  Automatic (A1)  2024  
34529              Model 3 Standard Range Plus       Y  Automatic (A1)  2019  
37234               Model Y Standard Range RWD       N  Automatic (A1)  2021  
38476                              Model Y RWD       N  Automatic (A1)  2022  
38472                              Model 3 RWD       N  Automatic (A1)  2022  
39794                              Model 3 RWD       Y  Automatic (A1)  2023

# 2.1  Different order for different columns

You may choose to sort according to multiple columns, where sorting can be ascending or descending for different columns. For example, in the below cell, city08 and highway08 follow descending order while year follows ascending order.

Also, ‘inplace’ is set to True to make the changes in the original DataFrame df1

In [18]:
df1.sort_values(by=['city08','highway08','year'], ascending=[False,False,True], inplace= True)
df1.head(10)

city08  cylinders     fuelType  highway08     id     make  \
40254     153        NaN  Electricity        127  46622  Hyundai   
40623     153        NaN  Electricity        127  46957  Hyundai   
40256     151        NaN  Electricity        120  46624  Hyundai   
40625     151        NaN  Electricity        120  46959  Hyundai   
37173     150        NaN  Electricity        133  43821    Tesla   
31226     150        NaN  Electricity        122  38431  Hyundai   
32569     150        NaN  Electricity        122  39641  Hyundai   
33393     150        NaN  Electricity        122  40384  Hyundai   
35477     148        NaN  Electricity        132  42278    Tesla   
35472     145        NaN  Electricity        121  42273  Hyundai   

                                         model mpgData           trany  year  
40254  Ioniq 6 Long range RWD (18 inch Wheels)       N  Automatic (A1)  2023  
40623  Ioniq 6 Long range RWD (18 inch Wheels)       N  Automatic (A1)  2024  
40256               Ioniq 6 Standard Range RWD       N  Automatic (A1)  2023  
40625               Ioniq 6 Standard Range RWD       N  Automatic (A1)  2024  
37173          Model 3 Standard Range Plus RWD       Y  Automatic (A1)  2021  
31226                           Ioniq Electric       N  Automatic (A1)  2017  
32569                           Ioniq Electric       N  Automatic (A1)  2018  
33393                           Ioniq Electric       N  Automatic (A1)  2019  
35477              Model 3 Standard Range Plus       N  Automatic (A1)  2020  
35472                           Ioniq Electric       N  Automatic (A1)  2020

# 2.2  Sorting based on index

Sorting can also be done on row or column labels using sort_index(). The below code shows sorting based on row labels. From the previous cell, we can see that df1, has been sorted and hence therow index is jumbled up. Let us put the row index back in order by sorting the DataFrame based on value of row indices.

In [6]:
df1.sort_index()

city08  cylinders fuelType  highway08     id        make  \
0          19        4.0  Regular         25      1  Alfa Romeo   
1           9       12.0  Regular         14     10     Ferrari   
2          23        4.0  Regular         33    100       Dodge   
3          10        8.0  Regular         12   1000       Dodge   
4          17        4.0  Premium         23  10000      Subaru   
...       ...        ...      ...        ...    ...         ...   
47070      19        4.0  Regular         26   9995      Subaru   
47071      20        4.0  Regular         28   9996      Subaru   
47072      18        4.0  Regular         24   9997      Subaru   
47073      18        4.0  Regular         24   9998      Subaru   
47074      16        4.0  Premium         21   9999      Subaru   

                     model mpgData            trany  year  
0       Spider Veloce 2000       Y     Manual 5-spd  1985  
1               Testarossa       N     Manual 5-spd  1985  
2                  Charger       Y     Manual 5-spd  1985  
3      B150/B250 Wagon 2WD       N  Automatic 3-spd  1985  
4         Legacy AWD Turbo       N     Manual 5-spd  1993  
...                    ...     ...              ...   ...  
47070               Legacy       N  Automatic 4-spd  1993  
47071               Legacy       N     Manual 5-spd  1993  
47072           Legacy AWD       Y  Automatic 4-spd  1993  
47073           Legacy AWD       Y     Manual 5-spd  1993  
47074     Legacy AWD Turbo       N  Automatic 4-spd  1993  

[47075 rows x 10 columns]

We now create a new DataFrame named new_index_df from df1 where the index is set to the values in column year.

In [7]:
new_index_df = df1.set_index('year')

In [8]:
new_index_df.head(10)

city08  cylinders fuelType  highway08     id        make  \
year                                                             
1985      19        4.0  Regular         25      1  Alfa Romeo   
1985       9       12.0  Regular         14     10     Ferrari   
1985      23        4.0  Regular         33    100       Dodge   
1985      10        8.0  Regular         12   1000       Dodge   
1993      17        4.0  Premium         23  10000      Subaru   
1993      21        4.0  Regular         24  10001      Subaru   
1993      22        4.0  Regular         29  10002      Subaru   
1993      23        4.0  Regular         26  10003      Toyota   
1993      23        4.0  Regular         31  10004      Toyota   
1993      23        4.0  Regular         30  10005      Toyota   

                    model mpgData            trany  
year                                                
1985   Spider Veloce 2000       Y     Manual 5-spd  
1985           Testarossa       N     Manual 5-spd  
1985              Charger       Y     Manual 5-spd  
1985  B150/B250 Wagon 2WD       N  Automatic 3-spd  
1993     Legacy AWD Turbo       N     Manual 5-spd  
1993               Loyale       N  Automatic 3-spd  
1993               Loyale       Y     Manual 5-spd  
1993              Corolla       Y  Automatic 3-spd  
1993              Corolla       Y     Manual 5-spd  
1993              Corolla       Y  Automatic 4-spd

Let us now sort the DataFrame by row indices in descending order. The rows for recent years will appear on the top

In [9]:
new_index_df.sort_index(ascending=False)

city08  cylinders fuelType  highway08     id       make  \
year                                                            
2024      38        4.0  Regular         38  46722    Hyundai   
2024      31        4.0  Regular         36  47240     Nissan   
2024      12       12.0  Premium         19  47238    Bentley   
2024      27        4.0  Regular         35  47237     Nissan   
2024      32        4.0  Regular         40  47236     Nissan   
...      ...        ...      ...        ...    ...        ...   
1984      19        4.0  Regular         21  27275      Dodge   
1984      18        4.0  Regular         20  27274      Dodge   
1984      17        4.0  Regular         20  27273     Nissan   
1984      25        4.0   Diesel         29  27272      Mazda   
1984      21        4.0  Regular         31  27088  Chevrolet   

                        model mpgData                             trany  
year                                                                     
2024       Tucson Hybrid Blue       N                 Automatic (AM-S6)  
2024                    Kicks       N  Automatic (variable gear ratios)  
2024              Flying Spur       N                 Automatic (AM-S8)  
2024                    Versa       N                      Manual 5-spd  
2024                    Versa       N  Automatic (variable gear ratios)  
...                       ...     ...                               ...  
1984  Power Ram 50 Pickup 4WD       N                      Manual 4-spd  
1984  Power Ram 50 Pickup 4WD       N                      Manual 4-spd  
1984               Pickup 4WD       N                      Manual 5-spd  
1984   B2000/B2200 Pickup 2WD       N                      Manual 5-spd  
1984           Cavalier Wagon       N                      Manual 5-spd  

[47075 rows x 9 columns]

2.3  Sorting based on column labels

Sorting can also work on column labels by setting the ‘axis’ parameter to 1.  The below cell will sort the DataFrame new_index_df based on column labels in descending order

In [11]:
new_index_df.sort_index(axis=1,ascending=False)

trany mpgData                model        make     id  \
year                                                                    
1985     Manual 5-spd       Y   Spider Veloce 2000  Alfa Romeo      1   
1985     Manual 5-spd       N           Testarossa     Ferrari     10   
1985     Manual 5-spd       Y              Charger       Dodge    100   
1985  Automatic 3-spd       N  B150/B250 Wagon 2WD       Dodge   1000   
1993     Manual 5-spd       N     Legacy AWD Turbo      Subaru  10000   
...               ...     ...                  ...         ...    ...   
1993  Automatic 4-spd       N               Legacy      Subaru   9995   
1993     Manual 5-spd       N               Legacy      Subaru   9996   
1993  Automatic 4-spd       Y           Legacy AWD      Subaru   9997   
1993     Manual 5-spd       Y           Legacy AWD      Subaru   9998   
1993  Automatic 4-spd       N     Legacy AWD Turbo      Subaru   9999   

      highway08 fuelType  cylinders  city08  
year                                         
1985         25  Regular        4.0      19  
1985         14  Regular       12.0       9  
1985         33  Regular        4.0      23  
1985         12  Regular        8.0      10  
1993         23  Premium        4.0      17  
...         ...      ...        ...     ...  
1993         26  Regular        4.0      19  
1993         28  Regular        4.0      20  
1993         24  Regular        4.0      18  
1993         24  Regular        4.0      18  
1993         21  Premium        4.0      16  

[47075 rows x 9 columns]

# 2.4  Sorting in the presence of missing values

The position of missing values in a sorted DataFrame can be controlled by using the ‘na_position’ parameter. When set to ‘first’, the NA values will appear on the top of the column which is being used for sorting.

In [12]:
df1.sort_values(by='cylinders', na_position= 'first',ascending=False)

city08  cylinders                    fuelType  highway08     id  \
7138       81        NaN                 Electricity         91  16423   
7139       81        NaN                 Electricity         64  16424   
8143       81        NaN                 Electricity         64  17328   
8144       74        NaN                 Electricity         58  17329   
8146       45        NaN                 Electricity         33  17330   
...       ...        ...                         ...        ...    ...   
32784      36        2.0  Premium Gas or Electricity         33  39835   
34308      30        2.0  Premium Gas or Electricity         31  41211   
32783      36        2.0  Premium Gas or Electricity         33  39834   
715        15        2.0                     Premium         23  10641   
27755      41        2.0  Premium Gas or Electricity         37  35279   

         make                           model mpgData           trany  year  
7138   Nissan                        Altra EV       N             NaN  2000  
7139   Toyota                         RAV4 EV       N             NaN  2000  
8143   Toyota                         RAV4 EV       N             NaN  2001  
8144     Ford                           Th!nk       N             NaN  2001  
8146     Ford          Explorer USPS Electric       N             NaN  2001  
...       ...                             ...     ...             ...   ...  
32784     BMW  i3s (94Ah) with Range Extender       N  Automatic (A1)  2018  
34308     BMW          i3 with Range Extender       N  Automatic (A1)  2019  
32783     BMW   i3 (94Ah) with Range Extender       N  Automatic (A1)  2018  
715     Mazda                            RX-7       Y    Manual 5-spd  1994  
27755     BMW                          i3 REX       Y  Automatic (A1)  2014  

[47075 rows x 10 columns]

When ‘na_position’ parameter is set to ‘last’, the missing values in the column used for sorting will appear at the end

In [13]:
df1.sort_values(by='cylinders', na_position='last')

city08  cylinders     fuelType  highway08     id        make  \
21115      15        2.0      Regular         22   2920       Mazda   
46696      16        2.0      Premium         22   9658       Mazda   
714        16        2.0      Premium         22  10640       Mazda   
715        15        2.0      Premium         23  10641       Mazda   
45501      15        2.0      Regular         23   8582       Mazda   
...       ...        ...          ...        ...    ...         ...   
41204     122        NaN  Electricity        104  47484  Volkswagen   
41205     108        NaN  Electricity         96  47485  Volkswagen   
41206     108        NaN  Electricity         96  47486  Volkswagen   
41207      79        NaN  Electricity         71  47487     Vinfast   
41208      71        NaN  Electricity         66  47488     Vinfast   

                model mpgData            trany  year  
21115            RX-7       N     Manual 5-spd  1987  
46696            RX-7       N  Automatic 4-spd  1993  
714              RX-7       N  Automatic 4-spd  1994  
715              RX-7       Y     Manual 5-spd  1994  
45501            RX-7       N     Manual 5-spd  1992  
...               ...     ...              ...   ...  
41204        ID.4 Pro     NaN   Automatic (A1)  2024  
41205    ID.4 AWD Pro     NaN   Automatic (A1)  2024  
41206  ID.4 AWD Pro S     NaN   Automatic (A1)  2024  
41207        VF 9 Eco     NaN   Automatic (A1)  2024  
41208       VF 9 Plus     NaN   Automatic (A1)  2024  

[47075 rows x 10 columns]

Congratulations on completing this demo.